File: tests-cypress/cypress/support/pages/login/login-page.ts
import { routes } from '../../constants/routes'

export const loginObj = {
  page: '[data-testid="login-page"]',
  form: {
    form: '[data-testid="login-form"]',
    username: '[data-testid="login-username"]',
    password: '[data-testid="login-password"]',
    rememberMe: '[data-testid="login-remember-me"]',
    submit: '[data-testid="login-submit"]',
  },
  errors: {
    server: '[data-testid="login-error"]',
    username: '[data-testid="login-username-error"]',
    password: '[data-testid="login-password-error"]',
  },
  hints: {
    slow: '[data-testid="login-slow-hint"]',
  },
}

export const loginPage = {
  visit: () => {
    cy.visit(routes.login)
    cy.get(loginObj.form.form).should('be.visible')
  },

  typeUsername: (value: string) => {
    cy.get(loginObj.form.username).clear().type(value)
  },

  typePassword: (value: string) => {
    cy.get(loginObj.form.password).clear().type(value, { log: false })
  },

  enableRememberMe: () => {
    cy.get(loginObj.form.rememberMe).check()
  },

  submit: () => {
    cy.get(loginObj.form.submit).click()
  },

  verifyShellReady: () => {
    cy.get(loginObj.form.username).should('be.visible')
    cy.get(loginObj.form.password).should('be.visible')
    cy.get(loginObj.form.submit).should('be.enabled')
  },

  verifyServerError: (message: string) => {
    cy.get(loginObj.errors.server).should('be.visible').and('contain.text', message)
  },

  verifyFieldErrors: () => {
    cy.get(loginObj.errors.username).should('be.visible')
    cy.get(loginObj.errors.password).should('be.visible')
  },

  verifyUsernameError: (message: string) => {
    cy.get(loginObj.errors.username).should('be.visible').and('contain.text', message)
  },

  verifyPasswordError: (message: string) => {
    cy.get(loginObj.errors.password).should('be.visible').and('contain.text', message)
  },

  verifyNoUsernameError: () => {
    cy.get(loginObj.errors.username).should('not.exist')
  },
}

File: tests-selenium/src/test/java/com/coinpulse/selenium/pages/LoginPage.java
package com.coinpulse.selenium.pages;

import com.coinpulse.selenium.constants.Routes;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * Login page object; selectors mirror the Cypress login page object.
 */
public class LoginPage extends BasePage {

    private static final By FORM = byTestId("login-form");
    private static final By USERNAME = byTestId("login-username");
    private static final By PASSWORD = byTestId("login-password");
    private static final By SUBMIT = byTestId("login-submit");
    private static final By SERVER_ERROR = byTestId("login-error");
    private static final By USERNAME_ERROR = byTestId("login-username-error");
    private static final By PASSWORD_ERROR = byTestId("login-password-error");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public void open() {
        openPath(Routes.LOGIN);
        awaitVisible(FORM);
    }

    public void awaitLoaded() {
        awaitVisible(FORM);
    }

    public void verifyShellReady() {
        awaitVisible(USERNAME);
        awaitVisible(PASSWORD);
        awaitEnabled(SUBMIT);
    }

    public void loginAs(String username, String password) {
        type(USERNAME, username);
        type(PASSWORD, password);
        click(SUBMIT);
    }

    public void typeUsername(String value) {
        type(USERNAME, value);
    }

    public void typePassword(String value) {
        type(PASSWORD, value);
    }

    public void submit() {
        click(SUBMIT);
    }

    public void verifyServerError(String message) {
        awaitVisible(SERVER_ERROR);
        awaitTextContains(SERVER_ERROR, message);
    }

    public void verifyFieldErrors() {
        awaitVisible(USERNAME_ERROR);
        awaitVisible(PASSWORD_ERROR);
    }

    public void verifyUsernameError(String message) {
        awaitVisible(USERNAME_ERROR);
        awaitTextContains(USERNAME_ERROR, message);
    }

    public void verifyPasswordError(String message) {
        awaitVisible(PASSWORD_ERROR);
        awaitTextContains(PASSWORD_ERROR, message);
    }

    public void verifyNoUsernameError() {
        awaitAbsent(USERNAME_ERROR);
    }
}
